docs/designs/data_transform.md

```diff
@@ -1,4 +1,84 @@
+standardize(age)
+normalize(size)
+categorize(city)
+
+In the terminology of TFX Transform, the word *analysis* refers to data statistics that derives the above parameters, given which, the *transforming* refers to the conversion of raw data into model inputs.  This design document is about making SQLFlow support data analysis, in addition to transformation.
+
+
+## The Syntax Extension
+
+We can ask users to write SQL statements for data analysis.  For example, we might want to assume that the following code snippet trains a model to predict the plate price from its size.
+
+```sql
+min, max = SELECT min(age), max(age) FROM plates;
+SELECT * FROM plates TO TRAIN DNNRegressor COLUMN normalize(size, min, max) LABEL price;
+```
+
+Unfortunately, SQL doesn't provide variables. The above example changes so much to the syntax and semantics that most SQL users cannot accept it.
+
+Also, to fulfill the goal of making deep learning more straightforward, we hope our users can write the following statement.
+
+```sql
+SELECT * FROM plates TO TRAIN DNNRegressor COLUMN normalize(size) LABEL price INTO a_model;
+```
+
+This syntax implies that SQLFlow can convert the above statement into a Python program that does data analysis and data transformation.
+
+To use the trained model to make predictions, we hope the users don't have to rewrite the COLUMN clause.   For example, the following statement assumes that the table `new_plates` has a field `size`. The converted Python code should normalize `size` before sending it to the input of `a_mode`.
+
+```sql
+SELECT * FROM new_plates TO PREDICT price USING a_model;
+```
+
+If the user doesn't want normalization, but the raw value, as the input, she could write the following.
+
+```sql
+SELECT * FROM new_plate TO PREDICT price USING a_model COLUMN size
+```
+
+Or, if she wants standardization instead of normalization, she could write the following.
+
+```sql
+SELECT * FROM new_plate TO PREDICT price USING a_model COLUMN standardize(size)
+```
+
+How if the table new_plate doesn't have the field used to train the model? How if there is no field named `size`, but a field `diameter`. We should allow users to use later as well.
+
+```sql
+SELECT * FROM new_plates TO PREDICT price USING a_model COLUM normalize(diameter)
+```
+
+Please be aware that if the user wants normalization, she must write `normalize(diameter)` instead of `diameter`.  The following example uses the raw value of `diameter` without normalization.
+
+```sql
+SELECT * FROM new_plates TO PREDICT price USING a_model COLUMN diameter
+```
+
+## The Challenge
+
+From the above examples, we see that a challenge is that the TO TRAIN clause must be able to save the input field names and the data analysis results and transformation steps together with the model, so to make sure that the predictions using the same data transformation as the training.
```

Comment:
Besides batch prediction job using `TO PREDICT`, we may also consider the scenario of model serving.

Reply:
I am afraid that it is too much detailed to mentioned multiple ways of serving. We can have another document on prediction services.